### PR TITLE
 [SPARK-27001][SQL] Refactor "serializerFor" method between ScalaReflection and JavaTypeInference

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This patch proposes refactoring `serializerFor` method between `ScalaReflection` and `JavaTypeInference`, being consistent with what we refactored for `deserializerFor` in #23854.

This patch also extracts the logic on recording walk type path since the logic is duplicated across `serializerFor` and `deserializerFor` with `ScalaReflection` and `JavaTypeInference`.

## How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+object WalkedTypePathRecorder {
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala]: I was thinking about making this as class (with storing built path in each instance) but soon realized it requires touching other thing as well and feel a bit overkill. I'm still open to make this as individual class so please let me know if it sounds better.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala]: making it a class looks better, as it needs to accumulate the walked type path.

**Comment by fottey** [comment]
As an outside observer, would this refactoring allow the method `ScalaReflection.serializeFor` to handle arbitrary types that conform to the Java bean interface, and/or common Java specific types, such as `java.util.List`?

I recently discovered that because most of the common Scala implicit encoders reduce to `ExpressionEncoder`'s `apply` method, it's very difficult to work with arbitrary Java bean type's in the Dataset API.  

Specifically, given a java bean type, `MyBean`, and an implicit encoder of that bean type in scope, existing Spark 2.4.0 machinery in can't synthesize a valid encoder at runtime for hybrid Scala / Java types, like `Seq[MyBean]` or tuple types like `(Int, MyBean)` despite the fact that we have encoders for `Seq[_]`, `Tuple2[_, _]`, and `MyBean` available separately.

While it may be unreasonable to solve the problem generically across all potential classes, it would be really nice if `ExpressionEncoder`'s `apply` method could somehow detect and support at least java beans and java.util.Lists at runtime... 

See below code examples:

```scala
import com.example.MyBean
import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder

object Example {
    case class Test()

    def main(args: Array[String]): Unit = {
        val spark: SparkSession = ???

        import spark.implicits._

        // Works today after above implicit import
        val ds: Dataset[Seq[Test]] = Seq(Seq(Test()), Seq(Test()), ...).toDS

        // DOES NOT WORK
        // ExpressionEncoder's apply method cannot handle type MyBean!
        implicit def newMyBeanExpressionEncoder: Encoder[MyBean] = ExpressionEncoder()
        // 
        // Need to do the following:
        implicit def newMyBeanBeanEncoder: Encoder[MyBean] = Encoders.bean(classOf[MyBean])

        // But this only allows expressing things like this:
        val ds: Dataset[MyBean] = Seq(new MyBean(), new MyBean(), ...).toDS

        // Due to the above limitation we CANNOT do the following, EVEN AFTER
        // newMyBeanBeanEncoder is brought into scope!
        // DOES NOT WORK 
        val ds: Dataset[Seq[MyBean]] = Seq(Seq(new MyBean()), Seq(new MyBean()), ...).toDS

        // Finally, these do not work: 

        // DOES NOT WORK 
        val ds: Dataset[(Int, MyBean)] = Seq((0, new MyBean()),(0, new MyBean()), ...).toDS

        // DOES NOT WORK
        implicit def newMyBeanEncoder: Encoder[Seq[MyBean]] = ExpressionEncoder()
        
        // DOES NOT WORK
        implicit def newMyBeanEncoder: Encoder[java.util.List[MyBean]] = ExpressionEncoder()

        // The above samples all rely on ExpressionEncoder
        // being able to handle every type in the expression...
        // currently seems to work for:
        // - case classes
        // - tuples
        // - scala.Product
        // - scala "primitives"
        // other common types with encoders... BUT NOT java beans... :'(
    }
}
```

**Comment by HeartSaVioR** [comment]
@fottey 
Could you file an issue for your case? Refactor is normally done without touching behavior, or at most making some minor changes, so I think it's out of scope for this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala]: Thanks for the support! Just addressed.

**Comment by HeartSaVioR** [comment]
Some test failures are occurred just because of one more new line. Fixed.
Some other test failures complain plan do not match: though string representation of two plans look same. Need to investigate.

**Comment by HeartSaVioR** [comment]
The default implementation of `equals` in WalkedTypePath affects comparison. Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala [hunk 2]
```diff
@@ -48,8 +48,8 @@ object DeserializerBuildHelper {
       expr: Expression,
       dataType: DataType,
       nullable: Boolean,
-      walkedTypePath: Seq[String],
-      funcForCreatingNewExpr: (Expression, Seq[String]) => Expression): Expression = {
+      walkedTypePath: WalkedTypePath,
+      funcForCreatingNewExpr: (Expression, WalkedTypePath) => Expression): Expression = {
     val newExpr = funcForCreatingNewExpr(expr, walkedTypePath)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: this belongs to the previous PR: why not just let the caller side create the expression and pass to `deserializerForWithNullSafety`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 3]
```diff
@@ -355,7 +352,7 @@ object JavaTypeInference {
    */
   def serializerFor(beanClass: Class[_]): Expression = {
     val inputObject = BoundReference(0, ObjectType(beanClass), nullable = true)
-    val nullSafeInput = AssertNotNull(inputObject, Seq("top level input bean"))
+    val nullSafeInput = AssertNotNull(inputObject, new WalkedTypePath(Seq("top level input bean")))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: why not use `recordRoot` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 4]
```diff
@@ -364,65 +361,35 @@ object JavaTypeInference {
     def toCatalystArray(input: Expression, elementType: TypeToken[_]): Expression = {
       val (dataType, nullable) = inferDataType(elementType)
       if (ScalaReflection.isNativeType(dataType)) {
-        NewInstance(
-          classOf[GenericArrayData],
-          input :: Nil,
-          dataType = ArrayType(dataType, nullable))
+        createSerializerForGenericArray(input, dataType, nullable = nullable)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: `createSerializerForPrimitiveArray`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+case class WalkedTypePath(walkedPaths: Seq[String] = Nil) extends Serializable {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: shall we use a mutable list for better performance?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: Yeah I might be considered too complicated. Not a big deal and looks simpler. Will address. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala (hunk 5 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: It was to address diverged paths for map key and value, but we can also copy instance via cloning internal list if necessary. Will address.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: Addressed via 90df8a3 - I found it a bit complicated to maintain the list without polluting, so please take a look at the change and let me know if you would like to roll back to immutable one if performance gain doesn't seem to have more value than complexity.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala [hunk 6]
```diff
@@ -39,40 +39,38 @@ object DeserializerBuildHelper {
       path: Expression,
       ordinal: Int,
       dataType: DataType,
-      walkedTypePath: Seq[String]): Expression = {
+      walkedTypePath: WalkedTypePath): Expression = {
     val newPath = GetStructField(path, ordinal)
     upCastToExpectedType(newPath, dataType, walkedTypePath)
   }
 
   def deserializerForWithNullSafety(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: so this is same as `expressionWithNullSafety`?

**Comment by viirya** [comment]
So this is a purely refactoring PR and doesn't address bug, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    Invoke(inputObject, "longValue", LongType)
+  }
+
+  def createSerializerForFloat(inputObject: Expression): Expression = {
+    Invoke(inputObject, "floatValue", FloatType)
+  }
+
+  def createSerializerForDouble(inputObject: Expression): Expression = {
+    Invoke(inputObject, "doubleValue", DoubleType)
+  }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala]: Seems these helper methods don't reduce code and just add one more wrapper around calling `Invoke`. Are they needed?

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala]: This is not for reducing code. This is for consistency. These methods ensure we are consistently serialize / deserialize things between ScalaReflection and JavaTypeInference if the type is same.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: Yeah now it's same. I've left comment on #23916. Btw I feel we might be better to focus review channel to one, either this PR or #23916 .

**Comment by HeartSaVioR** [comment]
> So this is a purely refactoring PR and doesn't address bug, right?

Yes, and make things consistent between `deserializerFor` and `serializerFor`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala (hunk 6 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: I didn't notice you also make a lot change to `DeserializerBuildHelper` in this PR.  There might be conflicts if continuing #23916, I will close it.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: I think `expressionWithNullSafety` is more general naming so might be preferred one, but `deserializerForWithNullSafety` is also a good name cause we have relevant method `deserializerForWithNullSafetyAndUpcast`.

So that's a matter of preference and either can be removed. Which method would we prefer to keep?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 8]
```diff
@@ -319,28 +323,26 @@ object ScalaReflection extends ScalaReflection {
...
+            deserializerFor(
+              fieldType,
+              addToPath(path, fieldName, dataType, newPathForField),
+              newPathForField)
+          }
           deserializerForWithNullSafety(
-            path,
+            newPath,
             dataType,
             nullable = nullable,
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Doesn't this revert what you did in previous PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: Thanks @viirya , please feel free to comment even it belongs to previous PR. Thanks again!

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: Why you remove the `funcForCreatingNewExpr` from this and turn to pass in created expression (deserializer)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 8 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: I got review comment for previous PR as well - this is the reflection of replacing function to applied expression which ends up making `deserializerForWithNullSafety` and `expressionWithNullSafety` being same.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: I tend to follow the suggestion (https://github.com/apache/spark/pull/23908#discussion_r261019295) unless I have strong opinion, as I'm fairly new to contributing SQL area. For consistency I agree having func is better, but for simplicity we can remove it like applying inline. Either is reasonable.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: I think we just need to keep `expressionWithNullSafety`. I don't see why we have to have 2 methods for `deserializeFor`. Leaving only a `deserializerForWithNullSafetyAndUpcast` is fine.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: Left `deserializerForWithNullSafetyAndUpcast` and `expressionWithNullSafety` since both are used in multiple places. Please let me know if it doesn't work.

**Comment by HeartSaVioR** [comment]
Rebased against latest master to incorporate #23913.

Btw, I've realized the boundary of refactor can be expanded to RowEncoder, for both `serializerFor` and `deserializerFor`. Would it be better to address it with new PR, or is it OK to address here as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import scala.collection.mutable
+
+case class WalkedTypePath() extends Serializable {
+  val walkedPaths: mutable.ArrayBuffer[String] = new mutable.ArrayBuffer[String]()
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: private?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import scala.collection.mutable
+
+case class WalkedTypePath() extends Serializable {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: We need to make this serializable? It seems this class is used only in a driver side?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import scala.collection.mutable
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: Probably, we would be better to describe what this class is used for.

**Comment by maropu** [comment]
Nice work, @HeartSaVioR! btw, this pr consists of the two parts you described in the PR description. If so, how about splitting this into the two prs for easy reviews? Refactoring the code for the consistency between ScalaReflection and JavaTypeInference, and adding WalkedTypePath then?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala [hunk 12]
```diff
@@ -39,40 +39,30 @@ object DeserializerBuildHelper {
...
+  def expressionWithNullSafety(
       expr: Expression,
       nullable: Boolean,
-      walkedTypePath: Seq[String]): Expression = {
+      walkedTypePath: WalkedTypePath): Expression = {
     if (nullable) {
       expr
     } else {
-      AssertNotNull(expr, walkedTypePath)
+      AssertNotNull(expr, walkedTypePath.copy())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: We can let `AssertNotNull` take a `Seq[String]`, to force us to copy the `WalkedTypePath` when creating `AssertNotNull`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala [hunk 13]
```diff
@@ -167,10 +157,10 @@ object DeserializerBuildHelper {
   private def upCastToExpectedType(
       expr: Expression,
       expected: DataType,
-      walkedTypePath: Seq[String]): Expression = expected match {
+      walkedTypePath: WalkedTypePath): Expression = expected match {
     case _: StructType => expr
     case _: ArrayType => expr
     case _: MapType => expr
-    case _ => UpCast(expr, expected, walkedTypePath)
+    case _ => UpCast(expr, expected, walkedTypePath.copy())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 14]
```diff
@@ -274,38 +274,40 @@ object JavaTypeInference {
...
 
       case _ if mapType.isAssignableFrom(typeToken) =>
         val (keyType, valueType) = mapKeyValueType(typeToken)
-        val newTypePath = (s"""- map key class: "${keyType.getType.getTypeName}"""" +
-          s""", value class: "${valueType.getType.getTypeName}"""") +: walkedTypePath
+        walkedTypePath.recordMap(keyType.getType.getTypeName,
+          valueType.getType.getTypeName)
+
+        val newTypePathForKey = walkedTypePath.copy()
+        val newTypePathForValue = walkedTypePath.copy()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: Sorry for the back and forth. But seems it's better to make `WalkedTypePath` immutable as there are branches. It's hard to maintain and we can easily mess it up if we forget the call `copy` somewhere.

**Comment by HeartSaVioR** [comment]
@maropu I like the idea of splitting PR, but since @cloud-fan already provides feedbacks on WalkedTypePath, might be better to hear opinion and decide. Let me first address his feedback on WalkedTypePath - even we decide to break down it would be needed work.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala (hunk 14 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: Yeah, same understanding. I'll revert back to let WalkedTypePath be immutable one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala (hunk 10 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: It was non-serializable and encountered error on existing tests so added it. Unfortunately I don't remember which one was.

**Comment by cloud-fan** [comment]
I'm OK both ways. Since the PR already contains the `WalkedTypePath` refactor, I think it's fine to include it as it's pretty intuitive.

**Comment by HeartSaVioR** [comment]
Thanks! I'll keep it as it is. How about applying `DeserializerBuildHelper`/`SerializerBuildHelper` to RowEncoder? Better to have it as separate PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 15]
```diff
@@ -387,38 +383,28 @@ object ScalaReflection extends ScalaReflection {
...
                     FloatType | DoubleType) =>
           val cls = input.dataType.asInstanceOf[ObjectType].cls
           if (cls.isArray && cls.getComponentType.isPrimitive) {
-            StaticInvoke(
-              classOf[UnsafeArrayData],
-              ArrayType(dt, false),
-              "fromPrimitiveArray",
-              input :: Nil,
-              returnNullable = false)
+            createSerializerForPrimitiveArray(input, dt)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: seems this branch is missing in the java side. We can address it in the followup.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 16]
```diff
@@ -1378,7 +1378,10 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
  * Cast the child expression to the target data type, but will throw error if the cast might
  * truncate, e.g. long -> int, timestamp -> data.
  */
-case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[String])
+case class UpCast(
+    child: Expression,
+    dataType: DataType,
+    walkedTypePath: WalkedTypePath = new WalkedTypePath())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: can we keep it `Seq[String]`? When we reach here, the `walkedTypePath` is only needed for logging/error message, and we don't need the `WalkedTypePath` class to help accumulate the paths.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 17]
```diff
@@ -1627,7 +1627,7 @@ case class InitializeJavaBean(beanInstance: Expression, setters: Map[String, Exp
  * `Int` field named `i`.  Expression `s.i` is nullable because `s` can be null.  However, for all
  * non-null `s`, `s.i` can't be null.
  */
-case class AssertNotNull(child: Expression, walkedTypePath: Seq[String] = Nil)
+case class AssertNotNull(child: Expression, walkedTypePath: WalkedTypePath = new WalkedTypePath())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 18]
```diff
@@ -358,7 +353,8 @@ object JavaTypeInference {
    */
   def serializerFor(beanClass: Class[_]): Expression = {
     val inputObject = BoundReference(0, ObjectType(beanClass), nullable = true)
-    val nullSafeInput = AssertNotNull(inputObject, Seq("top level input bean"))
+    val nullSafeInput = AssertNotNull(inputObject,
+      WalkedTypePath().recordRoot("top level input bean").getPaths)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: nit: we can keep it unchanged.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 19]
```diff
@@ -195,7 +195,7 @@ case class ExpressionEncoder[T](
         case r: BoundReference =>
           // For input object of Product type, we can't encode it to row if it's null, as Spark SQL
           // doesn't allow top-level row to be null, only its columns can be null.
-          AssertNotNull(r, Seq("top level Product or row object"))
+          AssertNotNull(r, WalkedTypePath().recordRoot("top level Product or row object").getPaths)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: nit: we can keep it unchanged.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 20]
```diff
@@ -1378,7 +1378,10 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
  * Cast the child expression to the target data type, but will throw error if the cast might
  * truncate, e.g. long -> int, timestamp -> data.
  */
-case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[String])
+case class UpCast(
+    child: Expression,
+    dataType: DataType,
+    walkedTypePath: Seq[String] = Nil)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: can we revert the code style change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 21]
```diff
@@ -28,7 +28,7 @@ import scala.util.Try
 import org.apache.spark.{SparkConf, SparkEnv}
 import org.apache.spark.serializer._
 import org.apache.spark.sql.Row
-import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow, ScalaReflection}
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow, ScalaReflection, WalkedTypePath}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala (hunk 18 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: Ah yes that's not even same. Will revert.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala (hunk 10 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala]: To check which test fails, I opened a pr as followup: #23973 (I'm a bit interested in this...).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 15 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Raised PR: https://github.com/apache/spark/pull/24015
